presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java

```diff
@@ -18,49 +18,148 @@
+        this.connectionProperties.setProperty("sendStringParametersAsUnicode", "false");
+        this.connectionProperties.setProperty("applicationName", "Presto");
+    }
+
+    @Override
+    public Set<String> getSchemaNames()
+    {
+        try (Connection connection = this.driver.connect(this.connectionUrl, this.connectionProperties);
+                ResultSet resultSet = connection.getMetaData().getSchemas()) {
+            ImmutableSet.Builder<String> schemas = ImmutableSet.builder();
+            while (resultSet.next()) {
+                String name = resultSet.getString("TABLE_SCHEM").toLowerCase(ENGLISH);
+                // skip internal schemas
+                if (!"sys".equals(name) && !"information_schema".equals(name)) {
+                    schemas.add(name);
+                }
+            }
+            return schemas.build();
+        } catch (SQLException e) {
+            throw new PrestoException(JDBC_ERROR, e);
+        }
     }
 
     @Override
     public void commitCreateTable(JdbcOutputTableHandle handle)
     {
         StringBuilder sql = new StringBuilder()
                 .append("sp_rename ")
-                .append(singleQuote(handle.getCatalogName(), handle.getSchemaName(), handle.getTemporaryTableName()))
+                .append("'" + quoted(handle.getCatalogName(), handle.getSchemaName(), handle.getTemporaryTableName()) + "'")
                 .append(", ")
-                .append(singleQuote(handle.getTableName()));
+                .append("'" + handle.getTableName() + "'");
 
-        try (Connection connection = getConnection(handle)) {
-            execute(connection, sql.toString());
+        try (Connection connection = this.getConnection(handle)) {
+            this.execute(connection, sql.toString());
         }
         catch (SQLException e) {
             throw new PrestoException(JDBC_ERROR, e);
         }
     }
 
-    private static String singleQuote(String catalog, String schema, String table)
+    @Override
+    public PreparedStatement getPreparedStatement(Connection connection, String sql)
+            throws SQLException
     {
-        return singleQuote(catalog + "." + schema + "." + table);
+        connection.setAutoCommit(false);
+        PreparedStatement statement = connection.prepareStatement(sql);
+        statement.setFetchSize(1000);
+        return statement;
     }
 
-    private static String singleQuote(String literal)
+    @Override
+    protected Type toPrestoType(int jdbcType, int columnSize)
+    {
+        if (jdbcType == Types.DATETIMEOFFSET) {
+            return TIMESTAMP_WITH_TIME_ZONE;
+        }
+        return super.toPrestoType(jdbcType, columnSize);
+    }
+
+    @Override
+    protected String toSqlType(Type type)
+    {
+        if (type instanceof VarcharType) {
+            if (((VarcharType) type).isUnbounded()) {
+                return "varchar(max)";
+            }
+            return "varchar(" + ((VarcharType) type).getLengthSafe() + ')';
```

Comment:
My assumptions were wrong, went back to read the doc, >8000 has to map to `varchar(max)`, I'll fix that.